src/ring_theory/non_unital_subsemiring/basic.lean

```diff
@@ -0,0 +1,820 @@
+
+@[mono]
+lemma to_add_submonoid_mono :
+  monotone (to_add_submonoid : non_unital_subsemiring R → add_submonoid R) :=
+to_add_submonoid_strict_mono.monotone
+
+/-- Construct a `non_unital_subsemiring R` from a set `s`, a subsemigroup `sg`, and an additive
+submonoid `sa` such that `x ∈ s ↔ x ∈ sg ↔ x ∈ sa`. -/
+protected def mk' (s : set R) (sg : subsemigroup R) (hg : ↑sg = s)
+  (sa : add_submonoid R) (ha : ↑sa = s) :
+  non_unital_subsemiring R :=
+{ carrier := s,
+  zero_mem' := ha ▸ sa.zero_mem,
+  add_mem' := λ x y, by simpa only [← ha] using sa.add_mem,
+  mul_mem' := λ x y, by simpa only [← hg] using sg.mul_mem }
+
+@[simp] lemma coe_mk' {s : set R} {sg : subsemigroup R} (hg : ↑sg = s)
+  {sa : add_submonoid R} (ha : ↑sa = s) :
+  (non_unital_subsemiring.mk' s sg hg sa ha : set R) = s := rfl
+
+@[simp] lemma mem_mk' {s : set R} {sg : subsemigroup R} (hg : ↑sg = s)
+  {sa : add_submonoid R} (ha : ↑sa = s) {x : R} :
+  x ∈ non_unital_subsemiring.mk' s sg hg sa ha ↔ x ∈ s :=
+iff.rfl
+
+@[simp] lemma mk'_to_subsemigroup {s : set R} {sg : subsemigroup R} (hg : ↑sg = s)
+  {sa : add_submonoid R} (ha : ↑sa = s) :
+  (non_unital_subsemiring.mk' s sg hg sa ha).to_subsemigroup = sg :=
+set_like.coe_injective hg.symm
+
+@[simp] lemma mk'_to_add_submonoid {s : set R} {sg : subsemigroup R} (hg : ↑sg = s)
+  {sa : add_submonoid R} (ha : ↑sa  =s) :
+  (non_unital_subsemiring.mk' s sg hg sa ha).to_add_submonoid = sa :=
+set_like.coe_injective ha.symm
+
+end non_unital_subsemiring
+
+namespace non_unital_subsemiring
+
+variables {F G : Type*} [non_unital_ring_hom_class F R S] [non_unital_ring_hom_class G S T]
+  (s : non_unital_subsemiring R)
+
+@[simp, norm_cast] lemma coe_zero : ((0 : s) : R) = (0 : R) := rfl
+@[simp, norm_cast] lemma coe_add (x y : s) : ((x + y : s) : R) = (x + y : R) := rfl
+@[simp, norm_cast] lemma coe_mul (x y : s) : ((x * y : s) : R) = (x * y : R) := rfl
+
+/-! Note: currently, there are no ordered versions of non-unital rings. -/
+
+@[simp] lemma mem_to_subsemigroup {s : non_unital_subsemiring R} {x : R} :
+  x ∈ s.to_subsemigroup ↔ x ∈ s := iff.rfl
+@[simp] lemma coe_to_subsemigroup (s : non_unital_subsemiring R) :
+  (s.to_subsemigroup : set R) = s := rfl
+@[simp] lemma mem_to_add_submonoid {s : non_unital_subsemiring R} {x : R} :
+  x ∈ s.to_add_submonoid ↔ x ∈ s := iff.rfl
+@[simp] lemma coe_to_add_submonoid (s : non_unital_subsemiring R) :
+  (s.to_add_submonoid : set R) = s := rfl
+
+/-- The non-unital subsemiring `R` of the non-unital semiring `R`. -/
+instance : has_top (non_unital_subsemiring R) :=
+⟨{ .. (⊤ : subsemigroup R), .. (⊤ : add_submonoid R) }⟩
+
+@[simp] lemma mem_top (x : R) : x ∈ (⊤ : non_unital_subsemiring R) := set.mem_univ x
+
+@[simp] lemma coe_top : ((⊤ : non_unital_subsemiring R) : set R) = set.univ := rfl
+
+/-- The preimage of a non-unital subsemiring along a non-unital ring homomorphism is a
+non-unital subsemiring. -/
+def comap (f : F) (s : non_unital_subsemiring S) : non_unital_subsemiring R :=
```

Comment:
Do any of the other `comap` definitions use `ring_hom_class` etc? I think we should either follow the established pattern of not using them, or first adapt the existing definitions to use them too (in order to learn whether this is actually a good idea).

Reply:
`submonoid.comap` does; `subgroup.comap`, `subring.comap` don't. I think this "established pattern" is primarily due to the fact that the morphism classes didn't exist until (semi)recently and certain subobject classes haven't yet been updated.